### PR TITLE
[docs] Add guide for code references

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Add guide for basic local code references functionality. Will stack PR for source control references, when ready.

## Test Plan

viewed locally

https://benpankow-guide-code-references.dagster.dagster-docs.io/guides/dagster/code-references

### CODE REVIEW

**Comment by benpankow** [comment]
* **#21755** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21755?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @benpankow and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: docs/content/_navigation.json
Comment (benpankow) [inline docs/content/_navigation.json:1162]: deprecated, so dropping

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:2]:
```suggestion
title: Linking to code in Dagster with Code References | Dagster Docs
```

Making this match the page title

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:5]: Should we mention assets in the title, since this seems to be specific to assets? Are we planning to expand this to other types of definitions?

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:12]: Thoughts on something like this to break things up a bit?

```suggestion
Using Code Reference metadata on your Dagster assets, you can easily view the asset's source code from the Dagster UI:

- **In local development**, navigate directly to the code backing an asset in your editor
- **In Dagster+**, link to source code in your source control repository to see the full change history for an asset

In this guide, we'll show you how to automatically and manually attach Code References to your Dagster assets.
```

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:27]: Because there's only one instructional section in this guide, I'd say we scrap `Referencing source code in your assets` and making the other two headings into H2s. If we plan to expand this to other definitions, though, maybe it's fine to keep it.

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:54]: Isn't this the **Asset details** page? Also, perhaps include the name of the field (**Code location**) so folks can immediately find it in the image

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:31]: I suggest we include the name of the file in this example so folks can make the connection between this and what is shown in the screen shot of the UI.

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:67]:
```suggestion
In some cases, you may want to manually attach code references to your asset definitions. For example, an asset whose definition is spread across multiple Python source files or is partially defined in YAML may have a more complex source structure.
```

For the YAML files - what does this refer to? Something like dbt models?

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:69]:
```suggestion
To manually attach code references to an asset definition, use an instance of `dagster._core.definitions.metadata.CodeReferencesMetadataSet`. You can then choose to augment these manual references with `with_source_code_references`:
```

Removing some redundancy

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:71]: Same suggestion here about including the name of the file somewhere

**Review by erinkcochran87 — COMMENTED** [review]
Looking good! One thing I'm not sure of is if it should be `Code References` or `code references` - is this a proprietary feature?

File: docs/content/guides/dagster/code-references.mdx
Comment (benpankow) [reply docs/content/guides/dagster/code-references.mdx:5]: Good call, updated!

File: docs/content/guides/dagster/code-references.mdx
Comment (benpankow) [reply docs/content/guides/dagster/code-references.mdx:67]: Updated to more clearly explain the example case.

File: docs/content/guides/dagster/code-references.mdx
Comment (benpankow) [reply docs/content/guides/dagster/code-references.mdx:33]: Good call, updated!

File: docs/content/guides/dagster/code-references.mdx
Comment (benpankow) [reply docs/content/guides/dagster/code-references.mdx]: Updated 👍

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:17]: You've got `Code References` here, and everywhere else it's `code references`. If this isn't a proprietary feature, lets go with `code references` to be consistent.

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:57]:
```suggestion
A link to the asset's source in `with_source_code_references.py` will then be visible in the **Asset Catalog** view in the Dagster UI:
```

pls format UI components (page names, buttons, copy, etc) in **bold**

File: docs/content/guides/dagster/code-references.mdx
Comment (erinkcochran87) [inline docs/content/guides/dagster/code-references.mdx:111]:
```suggestion
Each of the code references to `manual_references.py` will be visible in the **Asset details** page in the Dagster UI:
```

Same here, nitpicking about formatting UI stuff ⛏️

**Review by erinkcochran87 — COMMENTED** [review]
I'll leave it to the other reviewers to stamp for accuracy, but this looks good to me! I left a few small comments, pls address before merging 🙏

File: docs/content/guides/dagster/code-references.mdx
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:36]: We should expose this publicly if we're going to include it in a guide

File: docs/content/guides/dagster/code-references.mdx
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:51]: taking away the comma at the end of this line would allow this snippet to compose more

File: docs/content/_navigation.json
Comment (sryza) [reply docs/content/_navigation.json:1162]: not a huge deal, but ideally would do this in a separate PR where we also drop the mdx and python files

File: docs/content/guides/dagster/code-references.mdx
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:12]:
```suggestion
Attaching code reference metadata to your Dagster asset definitions allows you to easily view those assets' source code from the Dagster UI:
```

File: docs/content/guides/dagster/code-references.mdx
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:15]: This isn't just for Dagster+, right? They can also do this in their OSS deployment.

File: docs/content/guides/dagster/code-references.mdx
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:25]: "repository" is mostly a deprecated term. Maybe just "a set of Dagster asset definitions"?

File: docs/content/guides/dagster/code-references.mdx
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:74]: I think we're not ready yet to talk about `MetadataSet`s publicly.  We should just recommend
```python
metadata={
    "dagster/code_references": CodeReferencesMetadataValue(

...
```

similar to what we do in the column lineage guide: https://docs.dagster.io/concepts/metadata-tags/asset-metadata/column-level-lineage#enabling-column-level-lineage

File: docs/content/guides/dagster/code-references.mdx
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:97]: This makes me think we should make `line_number` optional

**Review by rexledesma — CHANGES_REQUESTED** [review]
Just clearing my queue -- let me know when this should get a final pass

**Comment by benpankow** [comment]
This is ready for review, yeah

File: docs/content/concepts/metadata-tags/asset-metadata.mdx
Comment (rexledesma) [inline docs/content/concepts/metadata-tags/asset-metadata.mdx]: Just a heads up to remove these merge conflicts

File: docs/content/guides/dagster/code-references.mdx
Comment (rexledesma) [inline docs/content/guides/dagster/code-references.mdx:2]: The possessive is a weird stylization in a title. I wonder if we can circumvent it altogether. We already use "asset definition" below anyways.

```suggestion
title: Linking to asset definition code with code references | Dagster Docs
```

File: docs/next/public/images/guides/code-references/manual_references.png
Comment (rexledesma) [inline docs/next/public/images/guides/code-references/manual_references.png:1]: I believe the designs for this have changed, [as mentioned here](https://dagsterlabs.slack.com/archives/C06AWMPT3DF/p1717620482045639?thread_ts=1717603935.833099&cid=C06AWMPT3DF). Has the updated frontend landed?

If not, I'm fine to approve, but we should update these images after the frontend updates.

**Review by rexledesma — APPROVED** [review]

**Comment by benpankow** [comment]
### Merge activity

* **Jul 8, 1:45 PM PDT**: @benpankow started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/21755).

**Pull request merged**
